accounting/src/delta/delta_data_collection/tests/delta_delta_delta_tests.rs

```diff
@@ -0,0 +1,107 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+#[rstest]
+#[rustfmt::skip]
+#[case(new_delta(None,      Some('a')), new_delta(Some('a'), Some('b')), new_delta(Some('b'), Some('c')), new_delta(None, Some('c')))]
+#[case(new_delta(None,      Some('a')), new_delta(Some('a'), Some('b')), new_delta(Some('b'), None),      new_delta(None, None))]
+#[case(new_delta(None,      Some('a')), new_delta(Some('a'), None),      new_delta(None,      Some('b')), new_delta(None, Some('b')))]
+#[case(new_delta(Some('a'), None),      new_delta(None,      Some('b')), new_delta(Some('b'), None),      new_delta(Some('a'), None))]
+#[case(new_delta(Some('a'), None),      new_delta(None,      Some('b')), new_delta(Some('b'), Some('c')), new_delta(Some('a'), Some('c')))]
+#[case(new_delta(Some('a'), Some('b')), new_delta(Some('b'), Some('c')), new_delta(Some('c'), None),      new_delta(Some('a'), None))]
+#[case(new_delta(Some('a'), Some('b')), new_delta(Some('b'), None),      new_delta(None, Some('c')),      new_delta(Some('a'), Some('c')))]
+fn delta_delta_delta_associativity(
+    #[case] delta1: DataDelta<char>,
+    #[case] delta2: DataDelta<char>,
+    #[case] delta3: DataDelta<char>,
+    #[case] expected_delta: DataDelta<char>,
+) {
+    let expected_collection = DeltaDataCollection::from_iter([(1, expected_delta)]);
+
+    {
+        // (Delta + Delta) + Delta = Delta
+        // every delta goes into separate collection
+        let mut collection1 = DeltaDataCollection::from_iter([(1, delta1.clone())]);
+        let collection2 = DeltaDataCollection::from_iter([(1, delta2.clone())]);
+        let collection3 = DeltaDataCollection::from_iter([(1, delta3.clone())]);
+        let _ = collection1.merge_delta_data(collection2).unwrap();
+        let _ = collection1.merge_delta_data(collection3).unwrap();
+
+        assert_eq!(collection1, expected_collection);
+    }
+
+    {
+        // Delta + (Delta + Delta) = Delta
+        // every delta goes into separate collection
+        let mut collection1 = DeltaDataCollection::from_iter([(1, delta1.clone())]);
+        let mut collection2 = DeltaDataCollection::from_iter([(1, delta2.clone())]);
+        let collection3 = DeltaDataCollection::from_iter([(1, delta3.clone())]);
+        let _ = collection2.merge_delta_data(collection3).unwrap();
+        let _ = collection1.merge_delta_data(collection2).unwrap();
+
+        assert_eq!(collection1, expected_collection);
+    }
+
+    {
+        // (Delta + Delta) + Delta = Delta
+        // every delta is applied to the same collection
+        let mut collection = DeltaDataCollection::new();
+        let _ = collection.merge_delta_data_element(1, delta1).unwrap();
+        let _ = collection.merge_delta_data_element(1, delta2).unwrap();
+        let _ = collection.merge_delta_data_element(1, delta3).unwrap();
+
+        assert_eq!(collection, expected_collection);
+    }
+}
+
```

Comment:
I think we can test the errors in a limited way too. If there is only one error, applying delta one by one should give the same error as combining deltas. Since errors only care about whether the value is `None` or `Some(_)`, we can use `Delta<()>` for testing. We have two deltas, first one applies cleanly, then the second one fails. Applying the second delta to the value updated by the first should give the same error as combining the two deltas together.

Here is the pseudocode:
```rust
fn error_consistency(x0, x1, x3: Delta<()>) {
    // Ensure x2 does not match
    let x2 = match x1 { None => Some(()), Some(()) => None };
    let d1 = Delta::new(x0, x1);
    let d2 = Delta::new(x2, x3);
    let apply_err = d2.apply(x1).expect_err("apply should fail");
    let combine_err = combine(d1, d2).expect_err("combine should fail");
    assert_eq!(apply_err, combine_err);
}
```

Then instantiate the test with all 8 combinations of `x0`, `x1`, `x3` = {`None`, `Some(())`}.

This does not give full associativity in presence of multiple errors but at least we can verify there is some consistency between updating a value using a delta and combining deltas together.

Reply:
This comment is probably attached to the wrong file but the idea still stands I think.

Reply:
Currently, errors for applying delta over data are more descriptive than for combining deltas. I agree that it makes sense to add such test, but it will compare the fact that an error is there (i.e. 'is_err()')

Reply:
> Currently, errors for applying delta over data are more descriptive than for combining deltas. I agree that it makes sense to add such test, but it will compare the fact that an error is there (i.e. 'is_err()')

Ok. I think it's fine to ignore this.